### PR TITLE
MM-35435 - add multiple SKU options to purchase modal

### PR DESCRIPTION
**Pull request by pablovelezvidal**
#### Summary
This PR adds the multiple sku options support to the purchase modal page. Now customers will be able to subscribe to a different cloud tier directly from the webapp.

#### Screenshots
Captura de pantalla 2021-05-14 a las 19 09 52

#### Release Note

```release-note
NONE
```

### CODE REVIEW

File: actions/cloud.tsx [hunk 1]
```diff
@@ -69,3 +74,14 @@ export function completeStripeAddPaymentMethod(stripe: Stripe, billingDetails: B
         return true;
     };
 }
+
+export function updateCloudSubscription(productId: string) {
```
Comment (marianunez) [inline actions/cloud.tsx:78]: To be consistent with the actual `client4` action

```suggestion
export function subscribeCloudSubscription(productId: string) {
```

File: components/common/radio_group.tsx [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+import React from 'react';
+
+type RadioGroupProps = {
+    id: string;
+    values: Array<{ key: string; value: string}>;
+    value: string;
+    isDisabled?: (id: string) => boolean | boolean;
+    onChange(e: React.ChangeEvent<HTMLInputElement>): void;
+}
+const RadioButtonGroup: React.FC<RadioGroupProps> = ({
```
Comment (marianunez) [inline components/common/radio_group.tsx:13]: Any tests we can add to this new component?

File: components/purchase_modal/purchase.scss [hunk 3]
```diff
@@ -72,6 +72,41 @@
                 margin-bottom: 24px;
                 min-width: 270px;
 
+                .select-plan {
+                    border-bottom: 1px solid rgba(0, 0, 0, 0.08);
```
Comment (marianunez) [inline components/purchase_modal/purchase.scss:76]: Should we be using theme colors here?

File: components/purchase_modal/purchase_modal.tsx [hunk 4]
```diff
@@ -44,17 +47,36 @@ type Props = {
...
+
+function findProductInDictionary(products: Dictionary<Product>, productId?: string): Product | null {
+    const keys = Object.keys(products);
+    let selectedProduct = null;
+    if (keys.length > 1) {
+        // here find the product by the provided id, otherwise return the one with Professional in the name
+        keys.forEach((key) => {
+            if (productId && products[key].id === productId) {
+                selectedProduct = products[key];
+            } else if (!productId && products[key].name.includes('Professional')) {
```
Comment (marianunez) [inline components/purchase_modal/purchase_modal.tsx:72]: If not found should we just fall back on the first one in the list (which should be Professional) so we are not married to a specific name

File: components/purchase_modal/purchase_modal.tsx [hunk 5]
```diff
@@ -44,17 +47,36 @@ type Props = {
...
+                selectedProduct = products[key];
+            }
+        });
+    } else if (keys.length === 1) {
+        selectedProduct = products[keys[0]];
+    }
```
Comment (marianunez) [inline components/purchase_modal/purchase_modal.tsx:78]: Related to the comment above, we should just do this if no selected product is found above.

File: components/purchase_modal/purchase_modal.tsx [hunk 6]
```diff
@@ -44,17 +47,36 @@ type Props = {
...
 type State = {
     paymentInfoIsValid: boolean;
-    productPrice: number;
     billingDetails: BillingDetails | null;
     cardInputComplete: boolean;
     processing: boolean;
+    selectedProduct: Product | null | undefined;
+}
+
+function findProductInDictionary(products: Dictionary<Product>, productId?: string): Product | null {
```
Comment (marianunez) [inline components/purchase_modal/purchase_modal.tsx:64]: Should this be a selector instead?

File: components/purchase_modal/purchase_modal.tsx [hunk 7]
```diff
@@ -64,26 +86,13 @@ export default class PurchaseModal extends React.PureComponent<Props, State> {
 
         this.state = {
             paymentInfoIsValid: false,
-            productPrice: 0,
             billingDetails: null,
             cardInputComplete: false,
             processing: false,
+            selectedProduct: findProductInDictionary(props.products!),
```
Comment (marianunez) [inline components/purchase_modal/purchase_modal.tsx:92]: If we don't expect `props.products` to be null then we should remove it from a possible type above in the type definition

File: components/purchase_modal/purchase_modal.tsx [hunk 8]
```diff
@@ -112,9 +121,57 @@ export default class PurchaseModal extends React.PureComponent<Props, State> {
         this.setState({processing: true, paymentInfoIsValid: false});
     }
 
+    comparePlan = (
+        <a
+            className='ml-1'
+            href='https://mattermost.com/pricing-cloud/'
```
Comment (marianunez) [inline components/purchase_modal/purchase_modal.tsx:127]: For compare plans like, please use: https://mattermost.com/pricing-cloud/#pricing-grid-block_5fa2028808529

My bad that I didn't include in the latest link in the Figma

File: components/purchase_modal/purchase_modal.tsx [hunk 9]
```diff
@@ -112,9 +121,57 @@ export default class PurchaseModal extends React.PureComponent<Props, State> {
...
+        >
+            <FormattedMessage
+                id='cloud_subscribe.contact_support'
+                defaultMessage='Compare plans'
+            />
+        </a>
+    );
+
+    onPlanSelected = (e: React.ChangeEvent<HTMLInputElement>) => {
+        const selectedPlan = findProductInDictionary(this.props.products!, e.target.value);
```
Comment (marianunez) [inline components/purchase_modal/purchase_modal.tsx:148]: Same comment as above regarding the use of the non-null assertion operator on `products`

**Review by marianunez — CHANGES_REQUESTED** [review]
Looking great @pablovelezvidal!

File: components/purchase_modal/purchase_modal.tsx [hunk 10]
```diff
@@ -36,54 +39,65 @@ let stripePromise: Promise<Stripe | null>;
...
+    const keys = Object.keys(products);
+    let selectedProduct = null;
+    if (keys.length > 1) {
+        // here find the product by the provided id, otherwise return the one with Professional in the name
+        keys.forEach((key) => {
+            if (productId && products[key].id === productId) {
+                selectedProduct = products[key];
+            } else if (!productId) {
+                selectedProduct = products[keys[0]];
+            }
```
Comment (marianunez) [inline components/purchase_modal/purchase_modal.tsx:78]: This check if no productId, doesn't seem related to this for each loop, should we check this at the beginning and return early?

File: components/purchase_modal/purchase_modal.tsx [hunk 11]
```diff
@@ -36,54 +39,65 @@ let stripePromise: Promise<Stripe | null>;
...
+}
+
+function findProductInDictionary(products: Dictionary<Product> | undefined, productId?: string): Product | null {
+    if (!products) {
+        return null;
+    }
+    const keys = Object.keys(products);
+    let selectedProduct = null;
+    if (keys.length > 1) {
+        // here find the product by the provided id, otherwise return the one with Professional in the name
```
Comment (marianunez) [inline components/purchase_modal/purchase_modal.tsx:72]: What happens if there if there is no keys? It seems that below it still tries to set the first element in the keys array. If there is no products, then it probably should exit early here too

File: components/threading/channel_threads/thread_footer/__snapshots__/thread_footer.test.tsx.snap [hunk 12]
```diff
@@ -712,7 +712,7 @@ exports[`components/threading/channel_threads/thread_footer should report total
               values={
                 Object {
                   "formatted": <Memo(SemanticTime)
-                    aria-label="Tue Sep 29 2020 02:30:15 GMT+0000"
+                    aria-label="Tue Sep 29 2020 04:30:15 GMT+0200"
```
Comment (marianunez) [inline components/threading/channel_threads/thread_footer/__snapshots__/thread_footer.test.tsx.snap:715]: Are these changes in this thread snap related to this PR?

File: components/admin_console/billing/billing_subscriptions/index.tsx [hunk 13]
```diff
@@ -70,6 +70,9 @@ const BillingSubscriptions: React.FC<Props> = () => {
         daysLeftOnTrial = getRemainingDaysFromFutureTimestamp(subscription.trial_end_at);
     }
 
+    isFreeTrial = true;
+    daysLeftOnTrial = 10;
```
Comment (marianunez) [inline components/admin_console/billing/billing_subscriptions/index.tsx:74]: We may need to remove this dev testing lines

**Review by marianunez — CHANGES_REQUESTED** [review]

File: components/purchase_modal/purchase_modal.tsx [hunk 14]
```diff
@@ -36,54 +39,69 @@ let stripePromise: Promise<Stripe | null>;
...
+    const keys = Object.keys(products);
+    if (!keys.length) {
+        return null;
+    }
+    if (!productId) {
+        return products[keys[0]];
+    }
+    let selectedProduct = null;
+    if (keys.length > 1) {
+        // here find the product by the provided id, otherwise return the one with Professional in the name
```
Comment (marianunez) [inline components/purchase_modal/purchase_modal.tsx:78]: Given that we already check for `key.length` above this can probably be removed as well and to simplify even further we can just initialized with the first one and replace in the loop if found. 
```suggestion
    let selectedProduct = products[keys[0]];;
    if (keys.length > 1) {
        // here find the product by the provided id, otherwise return the one with Professional in the name
```

File: components/purchase_modal/purchase_modal.tsx [hunk 15]
```diff
@@ -36,54 +39,69 @@ let stripePromise: Promise<Stripe | null>;
...
+        keys.forEach((key) => {
+            if (productId && products[key].id === productId) {
+                selectedProduct = products[key];
+            }
+        });
+    }
+
+    if (!selectedProduct) {
+        selectedProduct = products[keys[0]];
+    }
```
Comment (marianunez) [inline components/purchase_modal/purchase_modal.tsx:88]: We may be able to remove this based on the above comment

**Review by marianunez — APPROVED** [review]
Great work @pablovelezvidal!

**Comment by annelieseklein** [comment]
Looks great @pablovelezvidal ! Just a few questions/comments:

1. What is the size of the plan selector card on the right hand side? In the design I have 270px, but it looks wider here.
2. The copy below the "Subscribe" button looks a bit darker than what is in the design. I have 72% opacity of the center channel text.
3. The color of the radio buttons here also looks darker than what I have in the design.
4. Here it says "Mattermost Cloud Starter....". In the design I have just "Cloud Starter" but I'm not sure if we should already be updating it to just say "Starter" based on the packaging changes coming June 16? @marianunez

**Comment by pablovelezvidal** [comment]
> Looks great @pablovelezvidal ! Just a few questions/comments:
> 
> 1. What is the size of the plan selector card on the right hand side? In the design I have 270px, but it looks wider here.
> 2. The copy below the "Subscribe" button looks a bit darker than what is in the design. I have 72% opacity of the center channel text.
> 3. The color of the radio buttons here also looks darker than what I have in the design.
> 4. Here it says "Mattermost Cloud Starter....". In the design I have just "Cloud Starter" but I'm not sure if we should already be updating it to just say "Starter" based on the packaging changes coming June 16? @marianunez

Thanks for the feedback @annelieseklein, here are my updates:

1. Changed in the latest commit.
2. Changed in the latest commit.
3. Changed in the latest commit. One thing to note, the result is the unique height/width that you can manipulate to a radio button by using CSS. This because radio buttons are native elements and their look and feel is specific to each browser and OS. So, the workaround to get the radio button styled has to stay the same size/height of the original button displayed by the browser. Also, check the attached image on how it will display on Firefox. On this round that is the best we can do to style those radio buttons.
4. thoughts @marianunez ?

Chrome:

Captura de pantalla 2021-05-24 a las 17 25 33


Safari:
Captura de pantalla 2021-05-24 a las 17 26 24


Firefox:
Captura de pantalla 2021-05-24 a las 17 54 23

**Comment by marianunez** [comment]
> 4\. Here it says "Mattermost Cloud Starter....". In the design I have just "Cloud Starter" but I'm not sure if we should already be updating it to just say "Starter" based on the packaging changes coming June 16? @marianunez

Yes, this name is coming from the products we define in Stripe. We will adjust the name to say "Cloud Starter" or "Starter" based on the final decision for this screen. This will be done in the backend and not related to this PR. Just let us know what is the final copy @annelieseklein

**Review by annelieseklein — APPROVED** [review]

**Comment by amyblais** [comment]
@nickmisasi Kind reminder for PR review.

**Review by nickmisasi — APPROVED** [review]

**Review by stevemudie — APPROVED** [review]
Will test post merge.

**Comment by amyblais** [comment]
@pablovelezvidal If this is for the next Cloud release, it will need to be cherry-picked to the ``cloud`` branch.

**Comment by marianunez** [comment]
@amyblais This would be for Cloud June 16th release so it will probably make it in the next FF
